### PR TITLE
Google Cloud Storage - writer

### PR DESCRIPTION
**Pull request by Agares**
# Pull request

## Related

* Related [docs PR](https://github.com/tremor-rs/tremor-www/pull/223)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

### CODE REVIEW

File: src/connectors/impls/gcs/writer.rs [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+#[async_trait::async_trait]
+impl Connector for GCSWriterConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let url = Url::<HttpsDefaults>::parse(&self.config.endpoint)?;
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:67]: this ideally would be checked in `build_cfg` already so we get an error as soon as possible

File: src/connectors/impls/gcs/writer.rs [hunk 2]
```diff
@@ -0,0 +1,682 @@
...
+        let sink = GCSWriterSink {
+            client: None,
+            url: self.url.clone(),
+            config: self.config.clone(),
+            buffers: ChunkedBuffer::new(self.config.buffer_size),
+            current_name: None,
+            current_session_url: None,
+        };
+
+        let addr = builder.spawn(sink, sink_context)?;
```
Comment (mfelsche) [inline src/connectors/impls/gcs/writer.rs:110]:
```suggestion
        builder.spawn(sink, sink_context).map(Some)
```

File: src/connectors/impls/gcs/writer.rs [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+        Ok(Some(addr))
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
```
Comment (mfelsche) [inline src/connectors/impls/gcs/writer.rs:119]: This reminds me a lot of https://docs.rs/bytes/latest/bytes/struct.BytesMut.html

File: src/connectors/impls/gcs/writer.rs [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+        self.data = Vec::from(self.data.get(bytes_to_remove..).ok_or(
+            ErrorKind::GoogleCloudStorageError("Not enough data in the buffer"),
+        )?);
+        self.buffer_start += bytes_to_remove;
+
+        Ok(())
+    }
+
+    pub fn read_current_block(&self) -> Option<&[u8]> {
+        self.data.get(..self.block_size)
```
Comment (mfelsche) [inline src/connectors/impls/gcs/writer.rs:149]: they way the data is used below (the `Body` impl clones the slice again), it might be more efficient to use https://doc.rust-lang.org/std/vec/struct.Vec.html#method.split_off and return self and assign the returned vec to `self.data` (don't forget to set the capacity to `self.block_size * 2` afterwards).

To avoid `split_off` panicking, you could use it like this `self.data.split_off(self.block_size.min(self.data.len())`

File: src/connectors/impls/gcs/writer.rs [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+                            break;
+                        }
+
+                        sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+                    }
+                }
+
+                if let Some(response) = response {
+                    if response.status().is_server_error() {
+                        return Err("Received server errors from Google Cloud Storage".into());
```
Comment (mfelsche) [inline src/connectors/impls/gcs/writer.rs:263]: does the response body contain any useful error payload?

File: src/connectors/impls/gcs/writer.rs [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+            self.start_upload_if_needed(meta, name).await?;
+
+            let serialized_data = serializer.serialize(value, event.ingest_ns)?;
+            for item in serialized_data {
+                self.buffers.write(&item);
+            }
+
+            if let Some(data) = self.buffers.read_current_block() {
+                let mut response = None;
+                for i in 0..3 {
```
Comment (mfelsche) [inline src/connectors/impls/gcs/writer.rs:215]: Do we really want to keep this retry logic? We don't do similar things e.g. with the `http_client` connector. If people want retries, they can put a WAL in between or make their events `transactional` with other means. What do you think?

**Review by mfelsche — COMMENTED** [review]
Good stuff! I am a bit hesitant if we should keep the request retry logic though and/or make it configurable?

File: src/connectors/impls/gcs/writer.rs [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+        }
+    }
+
+    pub fn mark_done_until(&mut self, position: usize) -> Result<()> {
+        if position < self.buffer_start {
+            return Err("Buffer was marked as done at index which is not in memory anymore".into());
+        }
+
+        let bytes_to_remove = position - self.buffer_start;
+        self.data = Vec::from(self.data.get(bytes_to_remove..).ok_or(
```
[the remark this answers is not in the document]
Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:140]: Yes, but this should never happen, so that would potentially hide bugs in other places.

File: src/connectors/impls/gcs/writer.rs (hunk 6 above)
Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:215]: We could remove it, though in my experience, the errors here happen quite often.

File: src/connectors/impls/gcs/writer.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mfelsche) [reply src/connectors/impls/gcs/writer.rs:140]: Ah got it, because we are tracking the current cursor position in the overall bytes here too. :heavy_check_mark:

File: src/connectors/impls/gcs/writer.rs (hunk 6 above)
Comment (mfelsche) [reply src/connectors/impls/gcs/writer.rs:215]: The one big problem i have with this is that if we sleep here waiting for a retry, we are blocking the whole `sink` task from making progress. The http_client spawns a task for every http request sent. Could we somehow do the same here, keep track of the tasks/joinhandles/futures and chain those futures, so the uploads actually happen in order?

Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:215]: hmm... that could work, we could also have a background task that deals with the communication with GCS, but this brings another issue - how do we deal with ACKs then? AFAIK the ACK has to happen as `on_event` returns, unless I'm missing something

Comment (mfelsche) [reply src/connectors/impls/gcs/writer.rs:215]: Here is an example from the http_client: 
* https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/http/client.rs#L164
* https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/http/client.rs#L373-L378

File: src/connectors/impls/gcs/writer.rs (hunk 4 above)
Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:149]: I'm afraid we can't do this, because there's no guarantee that google will persist all of the data that we're sending to it. I.e. we might send 256kiB, but google can persists only 128kiB of that (or 0), so we need to keep it in memory until we receive a response and know how much we can throw away. 

```
If the request succeeds, the server responds with 308 Resume Incomplete. The response contains a [Range header](https://cloud.google.com/storage/docs/json_api/v1/parameters#range). Use the upper value in this header to determine where to start the next chunk. You should not assume that the server received all bytes sent in the request.
```
(https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload)

Comment (mfelsche) [reply src/connectors/impls/gcs/writer.rs:149]: Got it, that's why we need to have the actual cutting and removal of old stuff separate. Good point! :)

File: src/connectors/impls/gcs/writer.rs [hunk 8]
```diff
@@ -0,0 +1,831 @@
...
+#[cfg(test)]
+use tests::MockHttpClient;
+use tremor_pipeline::{ConfigImpl, Event};
+use tremor_value::Value;
+use value_trait::ValueAccess;
+
+#[derive(Deserialize, Debug, Clone)]
+pub struct Config {
+    #[serde(default = "default_endpoint")]
+    endpoint: String,
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:47]: we generalized to `url`  for everything that takes  url I think we should do it here too and not use `String` but `Url::<HttpsDefaults>` as a type.

File: src/connectors/impls/gcs/writer.rs [hunk 9]
```diff
@@ -0,0 +1,831 @@
...
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let default_bucket = self
+            .config
+            .bucket
+            .as_ref()
+            .map(|bucket_name| Value::from(bucket_name.as_str()).into_static());
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:114]:
```suggestion
            .cloned()
            .map(Value::from);
```

File: src/connectors/impls/gcs/writer.rs (hunk 8 above)
Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:47]: ohhh, right, I didn't think to even try deserializing this, but it's a good idea, will make the code a bit simpler :)

File: src/connectors/impls/gcs/writer.rs [hunk 10]
```diff
@@ -0,0 +1,831 @@
...
+
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct ChunkedBuffer {
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:133]: This could need a bit of documentation on what it does and what the intent is, I've a hard time figuring that out just from the code

File: src/connectors/impls/gcs/writer.rs [hunk 11]
```diff
@@ -0,0 +1,831 @@
...
+    }
+
+    pub fn final_block(&self) -> &[u8] {
+        self.data.as_slice()
+    }
+}
+
+enum HttpTaskCommand {
+    FinishUpload {
+        name: String,
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:185]: We use `bucket` in `StartUpload` so could this collide with having files with the same name but different buckets?

Comment (Licenser) [reply src/connectors/impls/gcs/writer.rs:185]: looking at the code below, yap that would clash

File: src/connectors/impls/gcs/writer.rs (hunk 6 above)
Comment (Licenser) [reply src/connectors/impls/gcs/writer.rs:339]: Since we have the retry logic in 3 places here, would it make sense to abstract it out, make it configurable and allow it be used elsewhere too?

hardcoded 3 tries at 25ms^2*i seems like a recepie for pain :joy:

File: src/connectors/impls/gcs/writer.rs [hunk 12]
```diff
@@ -0,0 +1,831 @@
...
+                    "The file name in metadata is not a string",
+                ))?;
+
+            self.finish_upload_if_needed(name).await?;
+
+            self.start_upload_if_needed(meta, name).await?;
+
+            let serialized_data = serializer.serialize(value, event.ingest_ns)?;
+            for item in serialized_data {
+                self.buffers.write(&item);
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:483]: the tasks seem to be bucket/name based, allowing for multiple names to be uploaded, but there seems to be only one buffer, I feel like I'm missing something here. are the bucket/name tasks simply "future-proofing" and not relevant in `on_event`?

Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:483]: The current way of working is that we're uploading all the events to a single buffer, until the name (and/or bucket) changes, and the `http_task` processes the events sequentially, so there's no need for multiple buffers.
I hope in the future we can support uploading multiple files in parallel, but this will require some careful analysis.

File: src/connectors/impls/gcs/writer.rs (hunk 6 above)
Comment (Agares) [reply src/connectors/impls/gcs/writer.rs:339]: Yep, all good points, I'll revamp the error handling and retries.

File: src/connectors/impls/gcs/writer.rs (hunk 12 above)
Comment (Licenser) [reply src/connectors/impls/gcs/writer.rs:483]: makes perfect sense!

File: src/connectors/impls/gcs/api_client.rs [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+        if let Some(mut response) = response {
+            if response.status().is_server_error() {
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+                return Err("Received server errors from Google Cloud Storage".into());
+            }
+
+            if let Some(raw_range) = response.header("range") {
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:143]: This is nicer written as:

```rust
let new_range = response.header("range").ok_or("No range header")?;
```

it removes a level of nesting and puts the error right next to where it occurs instead of 20 lines below :)

File: src/connectors/impls/gcs/api_client.rs [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+                    )
+                    .ok_or(ErrorKind::GoogleCloudStorageError(
+                        "Unable to get the end of the Range",
+                    ))?;
+
+                // NOTE: The data can only be thrown away to the point of the end of the Range header,
+                // since google can persist less than we send in our request.
+                //
+                // see: https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+                done_until.store(range_end.parse()?, Ordering::Release);
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:169]: Is done_until really shared over multiple threads (do we require an Atomic here) if not, is the ordering the right one? A comment on the sharing logic explaining it would be helpful to avoid confusion in a few months.

File: src/connectors/impls/gcs/api_client.rs [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+        Ok(())
+    }
+
+    async fn start_upload(&mut self, url: &Url<HttpsDefaults>, file: FileId) -> Result<()> {
+        for i in 0..3 {
+            let url = url::Url::parse(&format!(
+                "{}/b/{}/o?name={}&uploadType=resumable",
+                url, file.bucket, file.name
+            ))?;
+            #[cfg(not(test))]
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:186]: can we isolate the `test` and `non test` code into a separate function that's swapped out? mixing `cfg(test)` into the code makes it extremely hard to follow.

File: src/connectors/impls/gcs/api_client.rs [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+            #[cfg(test)]
+            let request = Request::new(Method::Post, url);
+            #[cfg(not(test))]
+            {
+                request.insert_header("Authorization", self.token.header_value()?.to_string());
+            }
+
+            let response = self.client.send(request).await;
+
+            if let Ok(mut response) = response {
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:197]: It would be nice to log the error if one occurred, right now, it seems to be quietly swallowed and retried.

File: src/connectors/impls/gcs/api_client.rs [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+                    data.start + data.len() - 1,
+                    data.start + data.len()
+                ),
+            );
+
+            request.set_body(data.data.clone());
+
+            let response = self.client.send(request).await;
+
+            if let Ok(mut response) = response {
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:257]: Same as above, it'd be nice not to swallow this error completely if something goes wrong.

File: src/connectors/impls/gcs/api_client.rs [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+                error!(
+                    "Error from Google Cloud Storage: {}",
+                    response.body_string().await?
+                );
+            }
+
+            sleep(Duration::from_millis(25u64 * 2u64.pow(i))).await;
+        }
+
+        Ok(())
```
Comment (Licenser) [inline src/connectors/impls/gcs/api_client.rs:270]: Same for the above functions.

If I read the code correctly, if we fail three times to upload then we silently succeed this seems wrong, shouldn't the default response be an Error

File: src/connectors/impls/gcs/chunked_buffer.rs [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+    pub fn write(&mut self, data: &[u8]) {
+        self.data.extend_from_slice(data);
+    }
+
+    pub fn start(&self) -> usize {
+        self.buffer_start
+    }
+
+    // TODO: Consume self instead?
```
Comment (Licenser) [inline src/connectors/impls/gcs/chunked_buffer.rs:79]: That would be nice, saves us a clone :D

File: src/connectors/impls/gcs/writer.rs [hunk 20]
```diff
@@ -0,0 +1,949 @@
...
+    endpoint: Url<HttpsDefaults>,
+    #[serde(default = "default_connect_timeout")]
+    connect_timeout: u64,
+    #[serde(default = "default_buffer_size")]
+    buffer_size: usize,
+    bucket: Option<String>,
+}
+
+#[allow(clippy::unwrap_used)]
+fn default_endpoint() -> Url<HttpsDefaults> {
```
Comment (Licenser) [inline src/connectors/impls/gcs/writer.rs:53]: Would be nice to add a test-case for this to guarantee if ever an error sneaks in here we find it during testing

**Review by Licenser — CHANGES_REQUESTED** [review]
A lot of the errors are plain strings and not wrapped into `ErrorKind::GoogleCloudStorageError` it would be good to unify this so people know where the error comes from.


I am a bit unsure about handling errors during writes, I think the way the connector acts right now we introduce the possibility for file corruption on write, let me give an example of the failure condition I have in mind:

- we have 10 events 0..9 that contain the chunks of the file
- we receive then I'm quick succession so that when event 10 arrives at the connector not all events have been written yet, say only 1 and 2 have been written
- now event 3 fails catastrophically (as in retries don't work)
- we send a CB fail event for 3 causing the upstrea system to resend everything from 3 onwards
- we write event 4 to 10 successfully
- the re-sends for 3-10 arrive, we now add them at the end of the file

This would lead to a corrupted file that tremor tells the user was uploaded successfully. Or I am missing the code that handles this condition :D it's a big PR could totally be.

I'm not sure how to best handle this the only way I see is not acknowledging any writes until the entire file is written, but that itself has problems for "endless streaming" use cases like appending logs. An alternative would be to make the connector as non-transactional and ignore all write failures turning it into a udp-like fire and forget one?

File: src/connectors/impls/gcs/api_client.rs (hunk 14 above)
Comment (Agares) [reply src/connectors/impls/gcs/api_client.rs:169]: it's shared between multiple tasks (i.e. the background task writes, and the sink task reads), AFAIK we can't make any guarantees about which core will it run on.
I'll add some comments though.

**Review by Licenser — APPROVED** [review]
🚀 weeeeh!
